### PR TITLE
feat(conditions): Adding support for config based conditions

### PR DESCRIPTION
**Pull request by jeyrschabu**
- support for config based conditions
- support for config based whitelisted clusters

### CODE REVIEW

File: orca-core/src/main/java/com/netflix/spinnaker/orca/conditions/ConditionConfigurationProperties.java [hunk 1]
```diff
@@ -29,7 +31,10 @@
   private boolean enabled = false;
   private Long backoffWaitMs = TimeUnit.MINUTES.toMillis(5);
   private Long waitTimeoutMs = TimeUnit.MINUTES.toMillis(120);
+  private List<String> clusters;
+  private List<String> conditions;
```
Comment (emjburns) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/conditions/ConditionConfigurationProperties.java]: What is the format of these conditions/clusters?

File: orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    when:
+    def result = conditionSupplier.getConditions(stage)
+
+    then:
+    result.size() == numberOfConditions
+
+    where:
+    ctx                 | clusters        | configuredConditions | numberOfConditions
+    [cluster: "foo"]    | []              | []                   | 0
```
Comment (emjburns) [inline orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy]: so the behavior when there are no clusters is nothing happens?

**Comment by emjburns** [comment]
I screwed up the github review dialog  (and meant to just comment) but basically I have some questions about how this is designed to be used.

File: orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy (hunk 2 above)
Comment (jeyrschabu) [reply orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy]: In this assertion, the resulting number of conditions is 0 because there aren't any configured conditions. Defining a list of clusters in the config is analogous to whitelisting. Let me know if this is unclear

File: orca-core/src/main/java/com/netflix/spinnaker/orca/conditions/ConditionConfigurationProperties.java (hunk 1 above)
Comment (emjburns) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/conditions/ConditionConfigurationProperties.java]: A string for a condition doesn't seem very resilient. Or is it just a message to print?

File: orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy (hunk 2 above)
Comment (emjburns) [reply orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy]: What happens if there are 0 clusters but 2 conditions?

File: orca-core/src/main/java/com/netflix/spinnaker/orca/conditions/ConditionConfigurationProperties.java (hunk 1 above)
Comment (jeyrschabu) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/conditions/ConditionConfigurationProperties.java]: This is just a message to print. Aimed to be used for testing.

File: orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    then:
+    result.size() == numberOfResultingConditions
+
+    where:
+    ctx                 | clusters        | activeConditions     | numberOfResultingConditions
+    [cluster: "foo"]    | []              | []                   | 0
+    [cluster: "foo"]    | ["foo", "bar"]  | []                   | 0
+    [cluster: "foo"]    | ["bar"]         | [ "c1", "c2"]        | 0
+    [cluster: "foo"]    | ["foo", "bar"]  | [ "c1", "c2"]        | 2
```
Comment (emjburns) [inline orca-core/src/test/groovy/com/netflix/spinnaker/orca/conditions/ConfigurationBackedConditionSupplierSpec.groovy]: can you add the test for zero clusters but two conditions?
